### ISSUE TITLE
Some issue in LINE

### ISSUE THREAD

**Issue by Tadelaide** [thunlp/OpenNE#41]
# self.h_e = tf.nn.l2_normalize(tf.nn.embedding_lookup(self.embeddings, self.h), 1)
        # self.t_e = tf.nn.l2_normalize(tf.nn.embedding_lookup(self.embeddings, self.t), 1)
        # self.t_e_context = tf.nn.l2_normalize(tf.nn.embedding_lookup(self.context_embeddings, self.t), 1)
        self.h_e = tf.nn.embedding_lookup(self.embeddings, self.h)
        self.t_e = tf.nn.embedding_lookup(self.embeddings, self.t)
        self.t_e_context = tf.nn.embedding_lookup(self.context_embeddings, self.t)
        self.second_loss = -tf.reduce_mean(tf.log_sigmoid(self.sign*tf.reduce_sum(tf.multiply(self.h_e, self.t_e_context), axis=1)))

I see those code in method LINE, and I am concerned about the same question why we cannot use l2_normalize in this step. Could you please give me some brief explanation about this point?

By the way, this is a really nice repository. Thank you for sharing your work.

**Comment by zzy14** [comment]
Thanks.

Basically, you can use l2_normalize in this step. According to the author's implement, I comment these code here.

**Comment by Tadelaide** [comment]
Thank you for your response so quickly.

However, I have try it right now and it give me an output like 'micro': 0.21136173767752714, 'macro': 0.048879544686962276.  When I comment the l2_normalize, it can shows 50% accuracy in Wiki dataset with 50% data for training.

My input like: python src/main.py--method line --label-file data/wiki/Wiki_category.txt --input data/wiki/Wiki_edgelist.txt --graph-format edgelist --output vec_all.txt --order 2 --epoch 50 --clf-ratio 0.5

I am really confused about this point, and it seem like that if I choose l2_normalize before training the model the gradient disappear.

**Comment by zzy14** [comment]
Interesting result. Actually, the norms of embeddings are not the same among different nodes. If you check out the output of LINE or Deepwalk, you will find that the norms of some nodes with high degrees are very small. It means that they connect to so many neighbors that the model cannot find significant representations for them.

When it comes to normalization during training, I think it will influence the model somehow. So, just forget it :)

**Comment by Tadelaide** [comment]
You are right.

Actually, I have tried add L2_normalize in two place in training process, which one is before calculating loss just like code you comment and another is after it computing the sum of the dot product. Both of them shows a result like above.

However, by your response, I realise that in this process I only do a L2_normalize for the node selected by every batch. It lead to that those nodes with high number links  will be normalised for many times.

It confused me for a week and really thank you for sharing=)
